src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,143 +103,211 @@ public class FromBase64Transform : ICryptoTransform
+            return value == 32 || ((uint)value - 9 <= (13 - 9)) ? true : false;
         }
 
-        private byte[] ConvertFromBase64(byte[] temp, int effectiveCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetOutputSize(int bytesToTransform, Span<byte> tmpBuffer)
         {
-            // Get the number of 4 bytes blocks to transform
-            int numBlocks = (effectiveCount + _inputIndex) / 4;
+            int outputSize = Base64.GetMaxDecodedFromUtf8Length(bytesToTransform);
 
-            byte[] transformBuffer = new byte[_inputIndex + effectiveCount];
-            Buffer.BlockCopy(_inputBuffer, 0, transformBuffer, 0, _inputIndex);
-            Buffer.BlockCopy(temp, 0, transformBuffer, _inputIndex, effectiveCount);
+            const byte padding = (byte)'=';
+            int len = tmpBuffer.Length;
 
-            _inputIndex = (effectiveCount + _inputIndex) % 4;
-            Buffer.BlockCopy(temp, effectiveCount - _inputIndex, _inputBuffer, 0, _inputIndex);
+            // In Base64 there are maximum 2 padding chars
+            if (tmpBuffer[len - 2] == padding) outputSize--;
+            if (tmpBuffer[len - 1] == padding) outputSize--;
 
-            char[] tempChar = Encoding.ASCII.GetChars(transformBuffer, 0, 4 * numBlocks);
-            byte[] tempBytes = Convert.FromBase64CharArray(tempChar, 0, 4 * numBlocks);
-            return tempBytes;
+            return outputSize;
         }
 
-        private byte[] DiscardWhiteSpaces(byte[] inputBuffer, int inputOffset, int inputCount)
+        private void ConvertFromBase64(Span<byte> tmpBuffer, Span<byte> outputBuffer, out int consumed, out int written)
         {
-            int i, iCount = 0;
-            for (i = 0; i < inputCount; i++)
+            int bytesToTransform = _inputIndex + tmpBuffer.Length;
+
+            // Common case for bytesToTransform = 4
+            byte[] transformBufferArray = null;
+            Span<byte> transformBuffer = stackalloc byte[4];
+            if (bytesToTransform > 4)
             {
-                if (char.IsWhiteSpace((char)inputBuffer[inputOffset + i])) iCount++;
+                transformBuffer = transformBufferArray = ArrayPool<byte>.Shared.Rent(bytesToTransform);
             }
 
-            // If there's nothing to do, leave early
-            if (iCount == 0 && inputOffset == 0 &&
-                inputCount == inputBuffer.Length)
+            transformBuffer = transformBuffer.Slice(0, bytesToTransform);
+
+            // Copy _inputBuffer to transformBuffer and append tmpBuffer
+            Debug.Assert(_inputIndex < _inputBuffer.Length);
+            _inputBuffer.AsSpan(0, _inputIndex).CopyTo(transformBuffer);
+            tmpBuffer.CopyTo(transformBuffer.Slice(_inputIndex));
+
+            // Save data that won't be transformed to _inputBuffer, so it can be transformed later
+            _inputIndex = bytesToTransform & 3;     // bit hack for % 4
```

Comment:
`Debug.Assert(bytesToTransform < 8)`, to show that we're not discarding data with the mod.

Reply:
`bytesToTransform` is allowed to be `>= 8`. Here it is `int & int`, so it won't discard any data in the int-domain.

Do you mean something else?

Reply:
I... feel like I thought this code path was assuming between one and two blocks but I'm not seeing why I thought that right now.  If it handles more than two then disregard.